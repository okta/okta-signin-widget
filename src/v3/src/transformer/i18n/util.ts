/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxTransaction } from '@okta/okta-auth-js';

import {
  getI18nKey,
  getI18NParams,
} from '../../../../v2/ion/i18nUtils';
import {
  FieldElement, TokenReplacement, UISchemaElement,
} from '../../types';
import { loc } from '../../util';

export const addTranslation = ({
  element,
  name,
  i18nKey,
  params,
  defaultValue = '',
  tokenReplacement,
  noTranslate,
}: {
  element: UISchemaElement;
  name: string;
  i18nKey: string;
  params?: any;
  defaultValue?: string;
  tokenReplacement?: TokenReplacement,
  noTranslate?: boolean,
}): void => {
  // TODO: change translations to required field with default value
  // eslint-disable-next-line no-param-reassign
  element.translations = element.translations || [];
  const useDefault = !i18nKey || (noTranslate && !!defaultValue);
  element.translations.push({
    name,
    i18nKey,
    value: useDefault ? defaultValue : loc(i18nKey, 'login', params, tokenReplacement),
    noTranslate,
  });
};

export const addLabelTranslationToFieldElement = (
  transaction: IdxTransaction,
  element: FieldElement,
): void => {
  const { nextStep } = transaction;
  const { relatesTo, name: stepName } = nextStep!;
  const authenticatorKey = relatesTo?.value?.key;
  const fieldName = element.options.inputMeta.name;
  const path = authenticatorKey
    ? `${stepName}.${authenticatorKey}.${fieldName}`
    : `${stepName}.${fieldName}`;
  const i18nKey = getI18nKey(path);
  const params = getI18NParams(nextStep, authenticatorKey);
  if (i18nKey || element.label) {
    const noTranslate = element.options.inputMeta.customLabel;
    addTranslation({
      element, name: 'label', i18nKey, params, defaultValue: element.label, noTranslate,
    });
  }
};

export const addOptionalLabelTranslationToFieldElement = (
  element: FieldElement,
): void => {
  if (element.options.inputMeta.required === false) {
    addTranslation({
      element, name: 'optionalLabel', i18nKey: 'oie.form.field.optional',
    });
  }
};
