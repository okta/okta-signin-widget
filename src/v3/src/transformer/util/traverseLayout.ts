/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  FormBag,
  StepperLayout,
  UISchemaElement,
  UISchemaLayout,
  UISchemaLayoutType,
} from '../../types';

type PredicateFn = (uischema: UISchemaElement) => boolean;
type CallbackFn = (uischema: UISchemaElement) => void;

type Options = {
  layout: FormBag['uischema'];
  predicate: PredicateFn;
  callback: CallbackFn;
};

export const traverseLayout = (options: Options) => {
  const fn = (
    layout: FormBag['uischema'],
    predicateFn: PredicateFn,
    callback: CallbackFn,
  ) => {
    layout.elements.forEach((element) => {
      if (predicateFn(element)) {
        callback(element);
      }

      const { type } = element;

      if (type === UISchemaLayoutType.STEPPER) {
        (element as StepperLayout).elements
          .forEach((el) => fn(el, predicateFn, callback));
        return;
      }

      if ([UISchemaLayoutType.HORIZONTAL, UISchemaLayoutType.VERTICAL]
        .includes(type as UISchemaLayoutType)) {
        fn(element as UISchemaLayout, predicateFn, callback);
      }
    });
  };

  fn(options.layout, options.predicate, options.callback);
};
