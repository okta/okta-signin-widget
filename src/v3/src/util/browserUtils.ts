/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

export const isAndroid = (): boolean => (
  // Windows Phone also contains "Android"
  /android/i.test(navigator.userAgent)
    && !/windows phone/i.test(navigator.userAgent)
);

export const isIOS = (): boolean => (
  // iOS detection from: http://stackoverflow.com/a/9039885/177710
  /iPad|iPhone|iPod/i.test(navigator.userAgent)
    // iPad on iOS 13 detection
    || (navigator.userAgent.includes('Mac') && 'ontouchend' in document)
);

export const isAndroidOrIOS = (): boolean => isAndroid() || isIOS();
