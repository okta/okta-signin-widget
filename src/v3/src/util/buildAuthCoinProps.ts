/*
 * Copyright (c) 2025-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxTransaction } from '@okta/okta-auth-js';

import {
  AUTHENTICATOR_KEY, DEVICE_ENROLLMENT_TYPE, EMAIL_AUTHENTICATOR_TERMINAL_KEYS, IDX_STEP,
} from '../constants';
import { AuthCoinProps } from '../types';
import { getAuthenticatorKey } from './getAuthenticatorKey';
import { containsOneOfMessageKeys } from './idx';

export const buildAuthCoinProps = (
  transaction?: IdxTransaction | null,
): AuthCoinProps | undefined => {
  if (!transaction) {
    return undefined;
  }

  const { nextStep, messages } = transaction;
  if (containsOneOfMessageKeys(EMAIL_AUTHENTICATOR_TERMINAL_KEYS, messages)
    || nextStep?.name === IDX_STEP.CONSENT_EMAIL_CHALLENGE) {
    return { authenticatorKey: AUTHENTICATOR_KEY.EMAIL };
  }

  if (nextStep?.name === IDX_STEP.PIV_IDP) {
    return { authenticatorKey: IDX_STEP.PIV_IDP };
  }

  if (nextStep?.name === IDX_STEP.REDIRECT_IDVERIFY && nextStep?.idp?.id) {
    return { authenticatorKey: nextStep.idp.id };
  }

  // @ts-expect-error Property 'deviceEnrollment' does not exist on type 'IdxContext' ts(2339)
  if (transaction.context?.deviceEnrollment?.value?.name === DEVICE_ENROLLMENT_TYPE.ODA) {
    return { authenticatorKey: AUTHENTICATOR_KEY.OV };
  }

  const authenticatorKey = getAuthenticatorKey(transaction);
  if (authenticatorKey) {
    return {
      authenticatorKey,
      // @ts-ignore logoUri missing from interface
      url: nextStep?.relatesTo?.value.logoUri,
    };
  }

  return undefined;
};
