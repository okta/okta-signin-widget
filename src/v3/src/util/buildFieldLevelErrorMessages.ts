/*
 * Copyright (c) 2024-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { WidgetMessage } from '../types';

type FieldLevelErrorMessages = {
  errorMessage?: string;
  errorMessageList?: string[];
};

export const buildFieldLevelErrorMessages = (
  errors: WidgetMessage[] | undefined,
) : FieldLevelErrorMessages => {
  const hasErrors = typeof errors !== 'undefined' && errors.length > 0;

  if (hasErrors) {
    if (errors.length === 1) {
      const error = errors[0];
      // Covers nested messages, e.g. password requirements not met
      // The SIW does not have messages nested more than once
      if (Array.isArray(error.message)) {
        return {
          errorMessage: error.description,
          errorMessageList: error.message
            .filter((msg) => typeof msg.message === 'string')
            .map((msg) => msg.message as string),
        };
      }
      return { errorMessage: error.message };
    }
    return {
      errorMessageList: errors
        .filter((error) => typeof error.message === 'string')
        .map((error) => error.message as string),
    };
  }

  return {};
};
