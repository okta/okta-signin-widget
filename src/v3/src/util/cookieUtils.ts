/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import Cookies from 'js-cookie';

const DEFAULT_EXPIRY_DAYS = 365;
const DEFAULT_COOKIE_PATH = '/';
const LAST_USERNAME_COOKIE_NAME = 'ln';
const USER_AUTHENTICATED_COOKIE_NAME = 'isuauth';

export const setCookieString = (
  name: string,
  value: string,
  expiryDays?: number,
  path?: string,
): void => {
  Cookies.set(
    name,
    value,
    { expires: expiryDays ?? DEFAULT_EXPIRY_DAYS, path: path ?? DEFAULT_COOKIE_PATH },
  );
};

export const getCookieString = (name: string): string => Cookies.get(name) as string;

export const getCookieUserAuthenticated = (): string => getCookieString(
  USER_AUTHENTICATED_COOKIE_NAME,
);

export const removeCookie = (name: string): void => Cookies.remove(name);

export const setUsernameCookie = (username: string): void => setCookieString(
  LAST_USERNAME_COOKIE_NAME,
  username,
);

export const setCookieUserAuthenticated = (): void => {
  Cookies.set(
    USER_AUTHENTICATED_COOKIE_NAME,
    '1',
    { expires: 1, path: DEFAULT_COOKIE_PATH },
  );
};

export const getUsernameCookie = (): string => getCookieString(LAST_USERNAME_COOKIE_NAME);

export const removeUsernameCookie = (): void => removeCookie(LAST_USERNAME_COOKIE_NAME);

export const removeUserAuthenticatedCookie = (): void => removeCookie(
  USER_AUTHENTICATED_COOKIE_NAME,
);
