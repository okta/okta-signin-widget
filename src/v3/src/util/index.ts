/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

export * from './browserUtils';
export * from './buildErrorMessageIds';
export * from './clipboard';
export * from './cookieUtils';
export * from './environmentUtils';
export * from './flattenInputs';
export * from './formatError';
export * from './formUtils';
export * from './generateRandomString';
export * from './getAuthenticatorKey';
export * from './getCurrentAuthenticator';
export * from './getElementKey';
export * from './getImmutableData';
export * from './getTranslation';
export * from './getValidationMessages';
export * from './idxUtils';
export * from './isInteractiveElement';
export * from './isPasswordRecovery';
export * from './isPollingStep';
export * from './languageUtils';
export * from './locUtil';
export * from './passwordUtils';
export * from './removeFieldLevelMessages';
export * from './resetMessagesToInputs';
export * from './settingsUtils';
export * from './shouldShowCancelLink';
export * from './toNestedObject';
export * from './webauthnUtils';
