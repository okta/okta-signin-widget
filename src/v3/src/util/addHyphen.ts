/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
*/
import { KeyboardEvent } from "src/types"

export const addHyphen = (e: KeyboardEvent<HTMLInputElement | HTMLTextAreaElement>): void => {
    const currentVal = e.currentTarget.value;
    // Add hyphen after 4th character. includes() check allows deletion of hyphen and prevents double hyphens
    if (currentVal && currentVal.length === 4 && !['Backspace', 'Delete', '-'].includes(e.key)) {
      e.currentTarget.value = currentVal.concat('-');
    } else if (currentVal && currentVal.length > 4 && currentVal.charAt(4) !== '-') {
      // If user types too fast, the above concatenation can be skipped. This retroactively adds a hyphen if it wasn't added
      e.currentTarget.value = currentVal.slice(0, 4) + '-' + currentVal.slice(4);
    }
}
