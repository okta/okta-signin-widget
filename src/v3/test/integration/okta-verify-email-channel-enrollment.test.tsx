/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { waitFor } from '@testing-library/preact';
import { createAuthJsPayloadArgs, setup } from './util';
import mockResponse from '../../src/mocks/response/idp/idx/credential/enroll/enroll-ov-email-channel.json';

describe('okta-verify-email-channel-enrollment', () => {
  it('should render email channel form and send correct payload', async () => {
    const {
      container, findByText, findByLabelText, user, authClient,
    } = await setup({ mockResponse });

    const headerEle = await findByText(/Set up Okta Verify via email link/);
    await waitFor(() => expect(headerEle).toHaveFocus());
    await findByText(/Make sure you can access the email on your mobile device./);
    const emailEl = await findByLabelText('Email') as HTMLInputElement;
    const submitBtn = await findByText(/Send me the setup link/);

    expect(container).toMatchSnapshot();

    await user.type(emailEl, 'tester@okta1.com');
    expect(emailEl.value).toEqual('tester@okta1.com');
    await user.click(submitBtn);
    expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
      ...createAuthJsPayloadArgs('POST', 'idp/idx/challenge/send', {
        email: 'tester@okta1.com',
      }),
    );
  });
});
