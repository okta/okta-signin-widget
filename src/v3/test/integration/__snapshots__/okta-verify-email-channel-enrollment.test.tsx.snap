// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`okta-verify-email-channel-enrollment should render email channel form and send correct payload 1`] = `
<div>
  <div
    class="MuiScopedCssBaseline-root emotion-0"
  >
    <main
      class="auth-container main-container MuiBox-root emotion-1"
      data-commit="b9bbc0140703c3fbf0e2e58920362e70"
      data-version="0.0.0"
      dir="ltr"
      id="okta-sign-in"
      lang="en"
    >
      <div
        class="MuiScopedCssBaseline-root emotion-2"
      >
        <div
          class="MuiBox-root emotion-3"
        >
          <div
            class="MuiBox-root emotion-4"
          >
            <div
              class="okta-sign-in-header auth-header authCoinSpacing MuiBox-root emotion-5"
            >
              <h1
                class="MuiTypography-root MuiTypography-h1 emotion-6"
              />
              <div
                aria-hidden="true"
                class="mfa-okta-verify MuiBox-root emotion-7"
                data-se="factor-beacon"
              >
                <svg
                  aria-labelledby="mfa-okta-verify"
                  fill="none"
                  height="2.85714286rem"
                  role="img"
                  viewBox="0 0 48 48"
                  width="2.85714286rem"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <title
                    id="mfa-okta-verify"
                  >
                    Okta Verify
                  </title>
                  <path
                    clip-rule="evenodd"
                    d="M32.872 22.853a8.947 8.947 0 1 1-2.288-4.91L24 24.498l-3.205-3.19c-.482-.48-1.322-.48-1.804 0a1.26 1.26 0 0 0-.374.898c0 .34.133.659.374.898l4.107 4.09c.24.239.561.371.902.371.336 0 .664-.136.902-.372L38.636 13.52a18.089 18.089 0 0 0-1.634-1.967A17.947 17.947 0 0 0 24 6C14.059 6 6 14.059 6 24s8.059 18 18 18 18-8.059 18-18c0-2.972-.722-5.775-1.998-8.245l-7.13 7.098Z"
                    fill="#00297A"
                    fill-rule="evenodd"
                  />
                </svg>
              </div>
            </div>
            <div
              class="MuiBox-root emotion-8"
            >
              <form
                class="o-form MuiBox-root emotion-9"
                data-se="o-form"
                novalidate=""
              >
                <div
                  class="MuiBox-root emotion-10"
                >
                  <div
                    class="MuiBox-root emotion-11"
                  >
                    <div
                      class="MuiBox-root emotion-12"
                      data-se="identifier-container"
                      title="tester2@test.com"
                    >
                      <div
                        class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault emotion-13"
                        data-se="identifier"
                        translate="no"
                      >
                        <div
                          class="MuiChip-icon MuiChip-iconMedium MuiChip-iconColorDefault MuiBox-root emotion-14"
                        >
                          <svg
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-15"
                            fill="none"
                            focusable="false"
                            role="img"
                            viewBox="0 0 24 24"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              clip-rule="evenodd"
                              d="M15.5 6.5a3.5 3.5 0 1 1-7 0 3.5 3.5 0 0 1 7 0Zm2 0a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0ZM5 23c0-2.357.694-4.363 1.886-5.762C8.064 15.856 9.782 15 12 15c2.215 0 3.934.862 5.113 2.25C18.307 18.652 19 20.66 19 23h2c0-2.722-.807-5.216-2.363-7.046C17.067 14.107 14.785 13 12 13c-2.782 0-5.064 1.097-6.636 2.941C3.806 17.77 3 20.263 3 23h2Z"
                              fill="currentColor"
                              fill-rule="evenodd"
                            />
                            <title>
                              User
                            </title>
                          </svg>
                        </div>
                        <span
                          class="MuiChip-label MuiChip-labelMedium emotion-16"
                        >
                          tester2@test.com
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    class="MuiBox-root emotion-11"
                  >
                    <div
                      class="MuiBox-root emotion-18"
                    >
                      <h2
                        class="MuiTypography-root MuiTypography-h4 emotion-19"
                        data-se="o-form-head"
                        id="enrollment-channel-data_Title_Set_up_Okta_Verify_via_email_link_okta_verify_aut2h3fft4y9pDPCS1d7_1"
                        tabindex="-1"
                      >
                        Set up Okta Verify via email link
                      </h2>
                    </div>
                  </div>
                  <div
                    class="MuiBox-root emotion-11"
                  >
                    <div
                      class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth emotion-21"
                    >
                      <label
                        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-sizeMedium MuiInputLabel-outlined emotion-22"
                        data-shrink="false"
                        for="email"
                        id="email-label"
                      >
                        <span>
                          Email
                        </span>
                      </label>
                      <div
                        class="MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl emotion-23"
                        required="true"
                      >
                        <input
                          aria-invalid="false"
                          aria-labelledby="email-label"
                          autocomplete="email"
                          class="MuiInputBase-input emotion-24"
                          data-se="email"
                          id="email"
                          inputmode="email"
                          name="email"
                          required=""
                          type="text"
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="MuiBox-root emotion-11"
                  >
                    <div
                      class="MuiBox-root emotion-18"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph emotion-27"
                        data-se="o-form-explain"
                        id="enrollment-channel-data_Description_Make_sure_you_can_access_the_email_on_your_mobile_device_okta_verify_aut2h3fft4y9pDPCS1d7_3"
                      >
                        Make sure you can access the email on your mobile device.
                      </p>
                    </div>
                  </div>
                  <div
                    class="MuiBox-root emotion-11"
                  >
                    <button
                      class="MuiButtonBase-root MuiButton-root MuiButton-primary MuiButton-primaryPrimary MuiButton-sizeMedium MuiButton-primarySizeMedium MuiButton-disableElevation MuiButton-fullWidth MuiButton-root MuiButton-primary MuiButton-primaryPrimary MuiButton-sizeMedium MuiButton-primarySizeMedium MuiButton-disableElevation MuiButton-fullWidth emotion-29"
                      data-se="save"
                      tabindex="0"
                      type="submit"
                    >
                      Send me the setup link
                    </button>
                  </div>
                  <div
                    class="MuiBox-root emotion-11"
                  >
                    <div
                      class="MuiBox-root emotion-1"
                    >
                      Or 
                      <a
                        class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways switch-channel-link emotion-32"
                        href="#"
                      >
                        try a different way
                      </a>
                       to set up Okta Verify.
                    </div>
                  </div>
                  <div
                    class="MuiBox-root emotion-11"
                  >
                    <a
                      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways emotion-32"
                      data-se="switchAuthenticator"
                      href=""
                    >
                      Return to authenticator list
                    </a>
                  </div>
                  <div
                    class="MuiBox-root emotion-11"
                  >
                    <a
                      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways emotion-32"
                      data-se="cancel"
                      href=""
                    >
                      Back to sign in
                    </a>
                  </div>
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    </main>
  </div>
</div>
`;
